### ISSUE TITLE
Benchmark harness fixtures silently drift from the notebooks they claim to mirror (nb00 via #2044, nb05 via #1989); the nb05 path no longer builds

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2096]
## Observation

`bench/scripts/benchmark_notebook_sampler.py` (and `benchmark_quickstart_sampler.py`) mirror the notebooks' models by **copying** them. Nothing enforces that the copy stays current, so when a notebook's model changes the harness silently keeps measuring the old one. Two independent instances exist right now, from two different commits.

**nb00 — diverged by #2044.** `36d7189cf` ("quickstart fits a star-forming galaxy — single dust screen, cheap NUTS from MAP", 2026-08-23) changed the quickstart's model:

```
-sfh={"type": "tsnorm", "all_params": FREE},
+sfh={"type": "dpl", "all_params": FREE, "age_gyr": 13.1},
-    "type": "two_component",
+    "type": "single_component",
```

`00_quickstart.py` today is **DPL SFH + single-component Calzetti + baked-in nebular, D = 6**. The harness's `_build_nb00` is **tsnorm + two-component Calzetti + nebular off, D = 7**. Four differences: SFH family, dust component count, nebular treatment, and dimension. Every nb00 row in `bench/reports/2026-08-17_*` measures a model the quickstart has not shipped since 2026-08-23.

**nb05 — diverged by #1989**, and separately the harness no longer builds at all. `benchmark_notebook_sampler.py` still passes the retired `dust=` group and names `law_bc="calzetti"` with no `law_diff`:

```python
        dust=builders.dust.two_component(
            ...
            law_bc="calzetti",
```

Both spellings have since been retired — `dust=` was split into peer `dust_attenuation=` / `dust_emission=` groups, and #1989 made naming one half of a per-screen law pair without the other an error. So the nb05 path raises at model build. The physics half of that divergence is filed separately.

## Why this went unnoticed

The two failures compound. Because the script raises at build, nobody could run it to discover that its models had drifted — and two separate agents repairing it independently made two different repairs, producing two contradictory NUTS baselines for the same notebook and seed before the cause was traced.

There is a knock-on in the published record: `bench/reports/2026-08-20_cuda_device_matrix.md` Finding 15 frames its result around "the notebooks" using a tsnorm SFH. #2044 landed three days after that report. The finding remains correct about the *harness fixtures* it measured, but the tsnorm premise now holds for one shipped notebook (nb05), not three.

## Ask

Some mechanism that makes drift loud instead of silent. Candidates, none of them measured:

1. **Import the model from the notebook** rather than duplicating it — the notebooks are `.py` under jupytext, so the builder can be factored into something both import.
2. **A test that asserts the harness model and the notebook model produce identical predictions** for a fixed parameter vector. Probably the cheapest, and it fails loudly on exactly this class of change.
3. **Record a checksum of the model spec in each report row**, so a stale row is visibly stale rather than quietly wrong.

Separately, and regardless of which is chosen: repair `benchmark_notebook_sampler.py` so it builds, in **one** place. Its being unrunnable is what allowed the drift to accumulate, and independent per-worktree repairs are how the contradictory baselines arose.

## Not asked for here

Whether nb00's or nb05's *current* model is the right one is a separate question — this issue is only about the harness silently claiming to mirror a notebook it no longer mirrors.

**Comment by suchethac** [comment]
## A second script has the same defect, and one more symptom

`bench/scripts/benchmark_quickstart_sampler.py` was not covered in the original report and is in the same state:

1. **It carries the same retired grammar** — the `dust=` group and `law_bc=` without `law_diff` — so it raises at model build for the same reason `benchmark_notebook_sampler.py` does.
2. **It also builds D = 6 against its own published D = 7.** That is a dimension mismatch against the numbers it produced, independent of the dust question.

Only `benchmark_notebook_sampler.py` has been repaired so far, and it was repaired twice, independently, in two different working trees — with two different dust-law choices. Those choices are not cosmetic: on nb05 seed 7 they move the shipped NUTS call from R-hat 1.0043 / 4 divergences / min ESS 88.1 to 1.1426 / 166 / 3.0.

This strengthens the case for the ask above: the repair is a **decision**, not a mechanical fix, so it wants to be made once and centrally rather than re-derived per user. Two independent repairs already produced two contradictory baselines for the same notebook and seed before the cause was traced.

Concretely, the scripts currently in scope:

| script | builds? | mirrors its notebook? |
|---|---|---|
| `benchmark_notebook_sampler.py` | no (retired grammar) | nb05 no (#1989), nb00 no (#2044), nb01 unchecked |
| `benchmark_quickstart_sampler.py` | no (retired grammar) | no — and D = 6 vs its own published D = 7 |

`nb01` has still not been checked against `notebooks/01_why_jax.py`; it is the one remaining fixture in this family whose status is unknown.

**Comment by suchethac** [comment]
## Resolved — plus three corrections to what I wrote above

Implemented on `fix/harness-notebook-parity` (`a7f6fd01d`). Three things in my earlier comments were wrong or incomplete.

### 1. I had the D=6/D=7 framing backwards

I wrote that `benchmark_quickstart_sampler.py` "builds D = 6 against its own published D = 7", implying the published numbers were suspect. They are not.

The 2026-08-17 quickstart report is **internally consistent**: D = 7 *and* its naming of `met_logzsol` as the L=160 row's worst-mixing parameter both require free metallicity. So the report is correct, and the **committed builder drifted away from it after publication** by losing its `met` group. The measurements stand; the code moved. Recorded as a dated correction block in the report rather than by touching any number.

### 2. nb00 has a fifth difference, and it is the load-bearing one

I listed four: SFH family, dust component count, nebular treatment, dimension. The fifth is the **SSP grid** — `_build_nb00` uses the bare-stellar `fsps_prsc_miles_chabrier`; today's quickstart uses the nebular-baked `prsc_miles_chabrier_wNE`.

That is the difference that makes `neb={"type": "ssp"}` mean anything. Building the live quickstart model against the bare grid would silently drop its nebular emission. Fifteen spec keys differ in total.

### 3. nb01 is clean — and *why* is the reusable part

`_build_nb01` and `notebooks/01_why_jax.py` are spec-identical, maximum relative difference **0.0** across all six bands. `git log -L` on the model block shows no commit ever touched its physics — only ruff formatting and a `load_ssp` path-resolver swap.

The reason is structural: nb01 delegates to `recipes.mock_recovery_minimal()`, so there is **one** definition that changes for both at once. nb05 and nb00 drifted because each spelled its model out twice. That is the cheapest available fix for any future fixture, and it is worth more than the guard.

Also, `ctl-jwst` carried an unenforced "mirrors the notebook exactly" claim in its docstring. It happened to be true; it is now a declaration a test can fail on.

## The mechanism

`tools/check_harness_parity.py` + `tests/contract/test_harness_notebook_parity.py`. Every fixture declares a `parity=` block — **no default, so a fixture added without one fails**:

- `mirrors` — executes the notebook's own `py:percent` cells up to its `SEDModel` definition, then requires both an identical canonical spec (params + bands + SSP identity) and identical predicted photometry at a fixed parameter vector.
- `historical` — must differ from its declared `anchor` in **exactly** its declared keys, with the chain terminating at a `mirrors` fixture.
- `standalone` — must state why.

**Intentional `*pre` fixtures are distinguished by anchoring, not by exemption** — `05pre → 05 → 05_fitting_photometry.py`. Repairing `05pre` into a duplicate of `05` fails ("no longer differs from its anchor"); letting it drift outside `differs_in` fails ("undeclared differences"). Both are tested with injected drift, including replaying #1989 itself.

Of the issue's three candidate approaches: option 1 (import the builder from the notebook) was rejected because the `SEDModel.build` call is the pedagogical centrepiece of these notebooks and hoisting it into a shared helper degrades teaching material to serve a benchmark — and `05pre`/`00pre` have no notebook left to import from. Its real value, *using the notebook's own code as the reference*, is retained by executing the notebook's prefix from outside. Option 3 (checksum in report rows) was rejected because a checksum is only read when someone reruns a benchmark, and the drift accumulated precisely because nobody could run these scripts — and a digest cannot tell a deliberate update from a corruption.

Option 2 was strengthened with the spec comparison because **predictions alone are blind to prior-range changes**: the `00`/`00pre` difference includes `met_logzsol` `U(-2.0, 0.2)` vs `U(-1.5, 0.3)`, which is prediction-identical at a fixed parameter vector.

`benchmark_quickstart_sampler.py` is repaired by **deleting its model** and importing the shared `NOTEBOOKS` registry — the "repair it once, centrally" half of this issue. A new `00now` fixture tracks the *live* quickstart; nothing in the tree tracked `00_quickstart.py` before, which is how #2044 passed unnoticed.

Guard cost ~100 s, no fits, running in the PR-gating contract tier rather than marked `slow` — a guard that never runs is not a guard.
